cloud/logstream.go

```diff
@@ -2,32 +2,35 @@ package cloud
 
 import (
 	"context"
-	"sync"
+	"io"
+	"sync/atomic"
 
 	"github.com/earthly/cloud-api/logstream"
 	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
 	"github.com/pkg/errors"
 	"golang.org/x/sync/errgroup"
 )
 
-func (c *client) StreamLogs(ctx context.Context, buildID string, deltasCh chan []*logstream.Delta) error {
+// Deltas is a type for iterating over logstream deltas.
+type Deltas interface {
+	Next(ctx context.Context) ([]*logstream.Delta, error)
+}
+
+func (c *client) StreamLogs(ctx context.Context, buildID string, deltas Deltas) error {
 	streamClient, err := c.logstream.StreamLogs(c.withAuth(ctx), grpc_retry.Disable())
 	if err != nil {
 		return errors.Wrap(err, "failed to create log stream client")
 	}
 	eg, ctx := errgroup.WithContext(ctx)
-	var mu sync.Mutex
-	finished := false
+	var finished int32
```

Comment:
Is the change to atomics necessary as part of the bug fix?

It feels more readable with just a mutex.

Reply:
Interesting, I find the atomics more readable. I mean, it's mildly annoying that the variable isn't a boolean in this section; but later in the file, a single `atomic.Store...` to my eyes reads cleaner than `mu.Lock(); finished = true; mu.Unlock()`. Mostly because it's a fairly uncomplicated part of the logic, and taking up three lines for it looks weird to me.

I'll make it an `atomic.Bool` instead and see if that's a reasonable compromise?

Reply:
Just my luck, `atomic.Bool` is new as of go 1.19 :sweat_smile: 

Hey @alexcb how do you feel about bumping our go version? AFAIK the `ast` submodule is already at 1.19.

Reply:
seems fine to me, especially if we already have a submodule using it.

Reply:
@vladaionescu this is now an `atomic.Bool`. Let me know what you think. I don't particularly mind going back to a mutex, it just is perhaps against my nature to use a whole mutex lock for a single line :sweat_smile: especially when the logic around the `finished` variable is mostly uninteresting.

Reply:
Sounds good to me!